## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [3]:
import seaborn as sns
import numpy as np
import pandas as pd


titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
357,0,2,female,38.0,0,0,13.0000,S,Second,woman,False,NaN,Southampton,no,True
877,0,3,male,19.0,0,0,7.8958,S,Third,man,True,NaN,Southampton,no,True
163,0,3,male,17.0,0,0,8.6625,S,Third,man,True,NaN,Southampton,no,True
818,0,3,male,43.0,0,0,6.4500,S,Third,man,True,NaN,Southampton,no,True
493,0,1,male,71.0,0,0,49.5042,C,First,man,True,NaN,Cherbourg,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [ ]:
print(titanic_data.isna().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64


### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [29]:
titanic_data = titanic_data.dropna(axis = "columns", thresh = len(titanic_data) // 2 + 1 )
titanic_data = titanic_data.dropna(axis = "index", thresh = len(titanic_data.columns) // 2 + 1)

### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [65]:
averAges = {"man": 0, "woman": 0, "child": 0}

for key in averAges.keys():
    averAges[key] = round(titanic_data.loc[titanic_data["who"] == key, "age"].median())

for key in averAges.keys():
    titanic_data.loc[titanic_data["who"] == key, "age"] = titanic_data.loc[titanic_data["who"] == key, "age"].fillna(averAges[key])

### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [50]:
titanic_data = titanic_data.dropna(thresh=len(titanic_data.columns) - 1)
print(titanic_data.isnull().sum().sum())

0


### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [64]:
print(titanic_data["embark_town"].value_counts().index[0])  

Southampton


### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [63]:
survived_percent = round(titanic_data["survived"].mean() * 100, 2)
print(survived_percent)

38.25


### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [62]:
print(titanic_data[titanic_data["survived"] == 1]["embark_town"].value_counts())


embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64


### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [61]:
survival_by_class = {}

for passenger_class in titanic_data["class"].unique():
    class_passengers = titanic_data[titanic_data["class"] == passenger_class]
    survival_by_class[passenger_class] = round(class_passengers["survived"].mean() * 100, 2)

print(pd.Series(survival_by_class))

Third     24.24
First     62.62
Second    47.28
dtype: float64


### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [67]:
rich_passengers = titanic_data[titanic_data["fare"] >= 100]
print(round(rich_passengers["survived"].mean() * 100, 2))

73.58


### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [68]:
children_alone = titanic_data[(titanic_data["who"] == "child") & (titanic_data["alone"] == True)]
print(len(children_alone))

6


Какие выводы вы можете сделать о выживших пассажирах Титаника? 

на тонущем титанике и гемофобы зафембоятся